ports/zephyr/modbluetooth_zephyr.c

```diff
@@ -404,6 +809,131 @@ int mp_bluetooth_gap_peripheral_connect_cancel(void) {
 
 #endif // MICROPY_PY_BLUETOOTH_ENABLE_CENTRAL_MODE
 
+// Note: modbluetooth UUIDs store their data in LE.
+static struct bt_uuid *create_zephyr_uuid(const mp_obj_bluetooth_uuid_t *uuid) {
+    struct bt_uuid *result = (struct bt_uuid *)m_new(union uuid_u, 1);
+    if (uuid->type == MP_BLUETOOTH_UUID_TYPE_16) {
+        bt_uuid_create(result, uuid->data, 2);
+    } else if (uuid->type == MP_BLUETOOTH_UUID_TYPE_32) {
+        bt_uuid_create(result, uuid->data, 4);
+    } else {    //  MP_BLUETOOTH_UUID_TYPE_128
+        bt_uuid_create(result, uuid->data, 16);
+    }
+    return result;
+}
+
+static void gatt_db_add(const struct bt_gatt_attr *pattern, struct bt_gatt_attr *attr, size_t user_data_len) {
+
+    const union uuid_u *u = CONTAINER_OF(pattern->uuid, union uuid_u, uuid);
+    size_t uuid_size = sizeof(u->u16);
+
+    if (u->uuid.type == BT_UUID_TYPE_32) {
+        uuid_size = sizeof(u->u32);
+    } else if (u->uuid.type == BT_UUID_TYPE_128) {
+        uuid_size = sizeof(u->u128);
+    }
+
+    memcpy(attr, pattern, sizeof(*attr));
+
+    // Store the UUID.
+    attr->uuid = (const struct bt_uuid *)m_new(union uuid_u, 1);
+    memcpy((void *)attr->uuid, &u->uuid, uuid_size);
+
+    // Copy user_data to the buffer.
+    if (user_data_len) {
+        attr->user_data = m_new(uint8_t, user_data_len);
+        memcpy(attr->user_data, pattern->user_data, user_data_len);
+    }
+}
+
+static void add_service(const struct bt_uuid *u, struct bt_gatt_attr *attr) {
+    union uuid_u *uuid = (union uuid_u *)u;
+
+    size_t uuid_size = sizeof(uuid->u16);
+
+    if (uuid->uuid.type == BT_UUID_TYPE_32) {
+        uuid_size = sizeof(uuid->u32);
+    } else if (uuid->uuid.type == BT_UUID_TYPE_128) {
+        uuid_size = sizeof(uuid->u128);
+    }
+
+    gatt_db_add(&(struct bt_gatt_attr)BT_GATT_PRIMARY_SERVICE(&uuid->uuid), attr, uuid_size);
+}
+
+static void add_characteristic(struct add_characteristic *ch, struct bt_gatt_attr *attr_chrc, struct bt_gatt_attr *attr_value) {
+    struct bt_gatt_chrc *chrc_data;
+
+    // Add Characteristic Declaration
+    gatt_db_add(&(struct bt_gatt_attr)
+        BT_GATT_ATTRIBUTE(BT_UUID_GATT_CHRC,
+            BT_GATT_PERM_READ,
+            bt_gatt_attr_read_chrc, NULL,
+            (&(struct bt_gatt_chrc) {})), attr_chrc, sizeof(*chrc_data));
+
+    // Allow prepare writes
+    ch->permissions |= BT_GATT_PERM_PREPARE_WRITE;
+
+    // Add Characteristic Value
+    gatt_db_add(&(struct bt_gatt_attr)
+        BT_GATT_ATTRIBUTE(ch->uuid,
+            ch->permissions & GATT_PERM_MASK,
+            mp_bt_zephyr_gatts_attr_read, mp_bt_zephyr_gatts_attr_write, NULL), attr_value, 0);
+
+    chrc_data = attr_chrc->user_data;
+    chrc_data->properties = ch->properties;
+    chrc_data->uuid = attr_value->uuid;
+}
+
+static void ccc_cfg_changed(const struct bt_gatt_attr *attr, uint16_t value) {
+    // TODO
```

Comment:
Is it possible to raise an exception here, eg `OSError(MP_EOPNOTSUPP)`?  Or will that raise through zephyr's BLE bindings in a bad way?

Reply:
No, that's not a good idea, I will better implement what's pending on that TODO.